vscode-dotty/src/tasty-decompiler.ts

```diff
@@ -0,0 +1,212 @@
+                  vscode.window.showErrorMessage("The class file related to this TASTy file could not be found.")
+                  break;
+                case ERROR_OTHER:
+                  vscode.window.showErrorMessage("A decompilation error has occurred.")
+                  break;
+                default:
+                  vscode.window.showErrorMessage("Unknown Error.")
+                  break;
+              }
+            })
+          }
+        })
+      )
+  }
+
+  dispose(): void {
+		this.disposables.forEach(d => d.dispose())
+		this.disposables = []
+  }
+
+  /**
+   * Request the TASTy in `textDocument` to be decompiled
+   */
+  private requestDecompile(textDocument: vscode.TextDocument): Promise<TastyDecompileResult> {
+    const requestParams = asTastyDecompileParams(textDocument)
+    const canceller = new CancellationTokenSource()
+    const token = canceller.token
+
+    return new Promise<TastyDecompileResult>(resolve => {
+      resolve(vscode.window.withProgress({
+        location: ProgressLocation.Notification,
+        title: "Decompiling"
+      }, () => this.client.sendRequest(TastyDecompileRequest.type, requestParams, token)
+      ))
+    }).then( decompileResult => {
+        canceller.dispose()
+        return decompileResult
+    });
+  }
+
+    /** Is this document a tasty file? */
+    private isTasty(document: vscode.TextDocument): boolean {
+      return vscode.languages.match(this.documentSelector, document) > 0
+    }
+
+}
+
+/**
+ * Provider of virtual, read-only, scala documents
+ */
+export class DecompiledDocumentProvider implements vscode.TextDocumentContentProvider {
+  static scheme = 'decompiled';
+
+  private _documents = new Map<string, string>();
+  private _subscriptions: vscode.Disposable;
+
+  constructor() {
+    // Don't keep closed documents in memory
+		this._subscriptions = vscode.workspace.onDidCloseTextDocument(doc => this._documents.delete(doc.uri.toString()));
+  }
+
+  dispose() {
+		this._subscriptions.dispose();
+		this._documents.clear();
+  }
+
+	provideTextDocumentContent(uri: vscode.Uri): string {
+    let document = this._documents.get(uri.toString());
+		if (document) {
+			return document;
+    } else {
+      return 'Failed to load result.';
+    }
+  }
+
+  /**
+   * Creates a new virtual document ready to be provided and opened.
+   *
+   * @param textDocument  The document containing the TASTy that was decompiled
+   * @param content       The source code provided by the language server
+   */
+  makeScalaDocument(textDocument: vscode.TextDocument, content: string): vscode.Uri  {
+    let scalaDocument = textDocument.uri.with({
+      scheme: DecompiledDocumentProvider.scheme,
+      path:  textDocument.uri.path.replace(".tasty", ".scala")
+    })
+    this._documents.set(scalaDocument.toString(), content);
+    return scalaDocument;
+  }
+}
+
+/**
+ * WebView used as container for preformatted TASTy trees
+ */
+class TastyTreeView {
```

Comment:
coding style issue: this should be two-space indent, no semicolons.

Reply:
We use 
```scala
/** WebView used as container for preformatted TASTy trees
 *  ABC
 *  DEF
 */
```
or
```scala
/** WebView used as container for preformatted TASTy trees */
```